File: nut_bolt.py
"""
Nuts & Bolts Problem (Lock & Key problem)

Given a set of n nuts of different sizes and n bolts of different sizes. There is a one-one mapping 
between nuts and bolts. Match nuts and bolts efficiently.
Constraint: Comparison of a nut to another nut or a bolt to another bolt is not allowed. It means nut 
can only be compared with bolt and bolt can only be compared with nut to see which one is bigger/smaller.

Other way of asking this problem is, given a box with locks and keys where one lock can be opened by one 
key in the box. We need to match the pair.
"""

#	Constraint: Comparison of a nut to another nut or a bolt to another bolt is not allowed. It means nut can only 
#	be compared with bolt and #bolt can only be compared with nut to see which one is bigger/smaller.
#	Other way of asking this problem is, given a box with locks and keys where one lock can be opened by one key
#	in the box. We need to match the pair.


def partition(array, left, right, pivot):

	i = left
	j = 0
	
	while j < (right+1):
		
		if array[j] < pivot:
			temp1 = array[i]
			array[i] = array[j]
			array[j] = temp1
			i = i + 1
		elif array[j] == pivot:
			temp1 = array[j]
			array[j] = array[right]
			array[right] = temp1
			j = j - 1
			
		j += 1
		
	temp2 = array[i]
	array[i] = array[right]
	array[right] = temp2
	
	return i


def nuts_bolts(nuts, bolts, low, high):

	if low < high:
	
		nuts_index = partition(nuts, low, high, bolts[high])
		
		bolt_index= partition(bolts, low, high, nuts[nuts_index])
		
		
		nuts_bolts(nuts, bolts, low, nuts_index - 1)
		nuts_bolts(nuts, bolts, nuts_index+1, high)


def main():

	nuts = [2,3,1,5,4]
	bolts = [1,4,5,3,2]
	
	nuts_bolts(nuts, bolts, 0, len(bolts)-1)
	print(nuts)
	print(bolts)
	
if __name__ == '__main__':
	main()

File: nut_bolt.c
#include<stdio.h>
#include<stdio.h>
#include<string.h>


int partition(int arr[], int low, int high, int pivot)
{
        int j, i = low;

        int temp1, temp2;
        
        for (j = low; j < high; j++)
        {
            if (arr[j] < pivot)
            {
                temp1 = arr[i];
                arr[i] = arr[j];
                arr[j] = temp1;
                i++;
            } 
            else if(arr[j] == pivot)
            {
                temp1 = arr[j];
                arr[j] = arr[high];
                arr[high] = temp1;
                j--;
            }
        }
        temp2 = arr[i];
        arr[i] = arr[high];
        arr[high] = temp2;
 
        // Return the partition index of an array based on the pivot 
        // element of other array.
        return i;
}

void matchPairs(int nuts[], int bolts[], int low, int high)
{
        if (low < high)
        {
            // Choose last intacter of bolts array for nuts partition.
            int pivot = partition(nuts, low, high, bolts[high]);
 
            // Now using the partition of nuts choose that for bolts
            // partition.
            partition(bolts, low, high, nuts[pivot]);
 
            // Recur for [low...pivot-1] & [pivot+1...high] for nuts and
            // bolts array.
            matchPairs(nuts, bolts, low, pivot-1);
            matchPairs(nuts, bolts, pivot+1, high);
        }
        
}


int main()
{
        // Nuts and bolts are represented as array of intacters
        int nuts[] = {2,3,1,4,5,6};
        int bolts[] = {5,6,3,1,2,4};
 
        matchPairs(nuts, bolts, 0, 5);
		
		int i;
        for(i=0; i<6; i++)
        {
        	printf("%d", nuts[i]);
        
        }
        printf("\n");
        for(i=0; i<6; i++)
        {
        	printf("%d", bolts[i]);
        
        }
		
		return 0;
}


